fixes issue #91

zen37 opened the pull request.

preslavmihaylov commented on issuetracker/pivotaltracker/pivotaltracker.go:
```diff
@@ -34,15 +35,15 @@ func (it *IssueTracker) taskURLFrom(taskID string) string {
-	if tokens[0] == "pivotaltracker.com" {
-		tokens = append([]string{"https:"}, tokens...)
-	}
+	pattern := regexp.MustCompile(`^(?P<scheme>^(https?:))//pivotaltracker.com/n/projects/(?P<project>[a-zA-Z0-9-_.]+)`)
```
the `/n` is optional in the above url

preslavmihaylov reviewed (changes requested):
Since these changes are a bit risky - they hit important integrations, could you add unit tests in `github/github_test.go` which verify these functions can correctly transform an input URL to the expected format.

zen37 commented:
> Since these changes are a bit risky - they hit important integrations, could you add unit tests in `github/github_test.go` which verify these functions can correctly transform an input URL to the expected format.

Sorry, I cannot find github_test.go

preslavmihaylov commented:
Yeah, that doesn't exist yet, you gotta create it with the very first unit test :)

preslavmihaylov commented on issuetracker/github/github_test.go:
```diff
@@ -0,0 +1,31 @@
+		want  string
+	}{
+		{"https://github.com/preslavmihaylov/todocheck", "https://api.github.com/repos/preslavmihaylov/todocheck/issues/"},
+		{"github.com/preslavmihaylov/todocheck/", "https://api.github.com/repos/preslavmihaylov/todocheck/issues/"},
```
Could you also add testcases for users with numbers e.g. `user123`, users with hyphen (`user-123`), also repos with numbers & hyphens & underscores

preslavmihaylov commented on issuetracker/github/github_test.go:
```diff
@@ -0,0 +1,31 @@
+		testname := fmt.Sprintf("%q", tt.input)
+		t.Run(testname, func(t *testing.T) {
+			it.Origin = tt.input
+			res := it.issueAPIOrigin()
```
could you make the test against `IssueURLFor` instead of this private function? The private functions are implementation details which are unstable & can change anytime. The public API is supposed to be stable & shouldn't change very often

zen37 replied:
done

preslavmihaylov commented on issuetracker/github/github_test.go:
```diff
@@ -0,0 +1,31 @@
+	"testing"
+)
+
+func Test_IssueTracker_issueAPIOrigin(t *testing.T) {
```
After changing the tested func name you'll have to test the testname as well.

Also, could you replicate this test in the other components as well? - pivotaltracker & gitlab

zen37 replied:
done

zen37 commented:
Preslav,

Did the changes some time ago, did you get an automatic message the PR has
been updated?

On another topic, a project with a few packages, and every package needs a
config.
What's better? A small config yaml file in each package or one large yaml
file in the root of the projects with all configs needed by the packages ?

thanks

On Thu, Dec 3, 2020 at 7:06 AM Preslav Mihaylov <notifications@github.com>
wrote:

> *@preslavmihaylov* requested changes on this pull request.
> ------------------------------
>
> In issuetracker/github/github_test.go
> <https://github.com/preslavmihaylov/todocheck/pull/118#discussion_r535309961>
> :
>
> > @@ -0,0 +1,31 @@
> +package github
> +
> +import (
> +	"fmt"
> +	"testing"
> +)
> +
> +func Test_IssueTracker_issueAPIOrigin(t *testing.T) {
> +	var tests = []struct {
> +		input string
> +		want  string
> +	}{
> +		{"https://github.com/preslavmihaylov/todocheck", "https://api.github.com/repos/preslavmihaylov/todocheck/issues/"},
> +		{"github.com/preslavmihaylov/todocheck/", "https://api.github.com/repos/preslavmihaylov/todocheck/issues/"},
>
> Could you also add testcases for users with numbers e.g. user123, users
> with hyphen (user-123), also repos with numbers & hyphens & underscores
> ------------------------------
>
> In issuetracker/github/github_test.go
> <https://github.com/preslavmihaylov/todocheck/pull/118#discussion_r535311167>
> :
>
> > +	var tests = []struct {
> +		input string
> +		want  string
> +	}{
> +		{"https://github.com/preslavmihaylov/todocheck", "https://api.github.com/repos/preslavmihaylov/todocheck/issues/"},
> +		{"github.com/preslavmihaylov/todocheck/", "https://api.github.com/repos/preslavmihaylov/todocheck/issues/"},
> +	}
> +
> +	for _, tt := range tests {
> +
> +		var it IssueTracker
> +
> +		testname := fmt.Sprintf("%q", tt.input)
> +		t.Run(testname, func(t *testing.T) {
> +			it.Origin = tt.input
> +			res := it.issueAPIOrigin()
>
> could you make the test against IssueURLFor instead of this private
> function? The private functions are implementation details which are
> unstable & can change anytime. The public API is supposed to be stable &
> shouldn't change very often
> ------------------------------
>
> In issuetracker/github/github_test.go
> <https://github.com/preslavmihaylov/todocheck/pull/118#discussion_r535312524>
> :
>
> > @@ -0,0 +1,31 @@
> +package github
> +
> +import (
> +	"fmt"
> +	"testing"
> +)
> +
> +func Test_IssueTracker_issueAPIOrigin(t *testing.T) {
>
> After changing the tested func name you'll have to test the testname as
> well.
>
> Also, could you replicate this test in the other components as well? -
> pivotaltracker & gitlab
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/preslavmihaylov/todocheck/pull/118#pullrequestreview-544044700>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEJAM7YQ7HOUIVWZOMCMA4TSS6SQPANCNFSM4UGEWEXA>
> .
>

preslavmihaylov commented on issuetracker/gitlab/gitlab.go:
```diff
@@ -33,14 +34,18 @@ func (it *IssueTracker) taskURLFrom(taskID string) string {
-	tokens := common.RemoveEmptyTokens(strings.Split(it.Origin, "/"))
-	if !strings.HasPrefix(tokens[0], "http:") && !strings.HasPrefix(tokens[0], "https:") {
-		tokens = append([]string{"https:"}, tokens...)
+	pattern := regexp.MustCompile(`^(?P<scheme>^(https?:))//gitlab.com/(?P<owner>[a-zA-Z0-9\-]+)/(?P<repo>[a-zA-Z0-9-_.]+)/?$`)
```
gitlab can also be hosted on private server instances.

Any hostname should be valid here, not only `gitlab.com`, e.g. `gitlab.myorg.com`

preslavmihaylov commented on issuetracker/pivotaltracker/pivotaltracker.go:
```diff
@@ -34,15 +35,21 @@ func (it *IssueTracker) taskURLFrom(taskID string) string {
+	if strings.Contains(origin, "/n/") {
+		pattern = regexp.MustCompile(`^(?P<scheme>^(https?:))//pivotaltracker.com/n/projects/(?P<project>[a-zA-Z0-9-_.]+)`)
+	} else {
+		pattern = regexp.MustCompile(`^(?P<scheme>^(https?:))//pivotaltracker.com/projects/(?P<project>[a-zA-Z0-9-_.]+)`)
```
You can avoid the `if` statement by making `n/` optional in your regex.

e.g.
```
...pivotaltracker.com/(n/)?/projects...
```

preslavmihaylov commented on issuetracker/pivotaltracker/pivotaltracker_test.go:
```diff
@@ -0,0 +1,34 @@
+package pivotaltracker
```
could you also make the changes for the `redmine` issue tracker?

zen37 replied:
Use regex here?

```
// IssueAPIOrigin returns the URL for redmine's issue-fetching API
func (it *IssueTracker) issueAPIOrigin() string {
	return fmt.Sprintf("%s/issues/", it.Origin)
}
```

preslavmihaylov replied:
Oh, you're right there is no point to make these changes there. It is simple enough

preslavmihaylov commented on issuetracker/pivotaltracker/pivotaltracker_test.go:
```diff
@@ -0,0 +1,34 @@
+	"testing"
+)
+
+func Test_IssueTracker_IssueURLFor(t *testing.T) {
```
Could you modify the `Makefile` at the root of the project to also run the unit tests after invoking `make test`?

zen37 commented:
Would the modification be like a loop to go through all the folders and run
the testing files from there?

On Sun, Dec 20, 2020 at 12:02 PM Preslav Mihaylov <notifications@github.com>
wrote:

> *@preslavmihaylov* requested changes on this pull request.
> ------------------------------
>
> In issuetracker/pivotaltracker/pivotaltracker_test.go
> <https://github.com/preslavmihaylov/todocheck/pull/118#discussion_r546429196>
> :
>
> > @@ -0,0 +1,34 @@
> +package pivotaltracker
> +
> +import (
> +	"fmt"
> +	"testing"
> +)
> +
> +func Test_IssueTracker_IssueURLFor(t *testing.T) {
>
> Could you modify the Makefile at the root of the project to also run the
> unit tests after invoking make test?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/preslavmihaylov/todocheck/pull/118#pullrequestreview-556057432>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEJAM7ZTY7KGOLXRZ3VZYG3SVZJ4ZANCNFSM4UGEWEXA>
> .
>

preslavmihaylov commented:
@zen37 no, I think running a `go test` command while excluding the `testing` folder should suffice.

sattellite commented:
Does it will work for self-hosted installations of gitlab/github? Regexp pattern contains public service addresses: api.github.com and gitlab.com. I think this changes will not working for domains of self-hosted installations.

preslavmihaylov commented on issuetracker/gitlab/gitlab.go:
```diff
@@ -33,14 +34,19 @@ func (it *IssueTracker) taskURLFrom(taskID string) string {
-	if !strings.HasPrefix(tokens[0], "http:") && !strings.HasPrefix(tokens[0], "https:") {
-		tokens = append([]string{"https:"}, tokens...)
+
+	pattern := regexp.MustCompile(`^(?P<scheme>^(https?:))//(?P<domain>gitlab.([a-zA-Z0-9\-]+)?(.)?com)/(?P<owner>[a-zA-Z0-9\-]+)/(?P<repo>[a-zA-Z0-9-_.]+)/?$`)
```
gitlab can be self-hosted. It should be able to match any host e.g. `myorg.issuetracker.com`

preslavmihaylov replied:
In this case, you expect someone to explicitly have `gitlab` in the hostname which can not always be the case.

zen37 commented:
"gitlab can be self-hosted. It should be able to match any host e.g.
myorg.issuetracker.com"

so

"myorg.com/PRESLAVmihaylov/project"

should become

"https://myorg.com/api/v4/projects/preslavmihaylov%2Fproject/issues/

?

On Thu, Dec 24, 2020 at 4:51 AM Preslav Mihaylov <notifications@github.com>
wrote:

> *@preslavmihaylov* requested changes on this pull request.
> ------------------------------
>
> In issuetracker/gitlab/gitlab.go
> <https://github.com/preslavmihaylov/todocheck/pull/118#discussion_r548521554>
> :
>
> >  func (it *IssueTracker) issueAPIOrigin() string {
> -	tokens := common.RemoveEmptyTokens(strings.Split(it.Origin, "/"))
> -	if !strings.HasPrefix(tokens[0], "http:") && !strings.HasPrefix(tokens[0], "https:") {
> -		tokens = append([]string{"https:"}, tokens...)
> +
> +	pattern := regexp.MustCompile(`^(?P<scheme>^(https?:))//(?P<domain>gitlab.([a-zA-Z0-9\-]+)?(.)?com)/(?P<owner>[a-zA-Z0-9\-]+)/(?P<repo>[a-zA-Z0-9-_.]+)/?$`)
>
> gitlab can be self-hosted. It should be able to match any host e.g.
> myorg.issuetracker.com
> ------------------------------
>
> In issuetracker/gitlab/gitlab.go
> <https://github.com/preslavmihaylov/todocheck/pull/118#discussion_r548521656>
> :
>
> >  func (it *IssueTracker) issueAPIOrigin() string {
> -	tokens := common.RemoveEmptyTokens(strings.Split(it.Origin, "/"))
> -	if !strings.HasPrefix(tokens[0], "http:") && !strings.HasPrefix(tokens[0], "https:") {
> -		tokens = append([]string{"https:"}, tokens...)
> +
> +	pattern := regexp.MustCompile(`^(?P<scheme>^(https?:))//(?P<domain>gitlab.([a-zA-Z0-9\-]+)?(.)?com)/(?P<owner>[a-zA-Z0-9\-]+)/(?P<repo>[a-zA-Z0-9-_.]+)/?$`)
>
> In this case, you expect someone to explicitly have gitlab in the
> hostname which can not always be the case.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/preslavmihaylov/todocheck/pull/118#pullrequestreview-558495018>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEJAM7536KBXACKNHVIL673SWM2MTANCNFSM4UGEWEXA>
> .
>

preslavmihaylov commented:
> "gitlab can be self-hosted. It should be able to match any host e.g. myorg.issuetracker.com" so "myorg.com/PRESLAVmihaylov/project" should become "https://myorg.com/api/v4/projects/preslavmihaylov%2Fproject/issues/ ?
> […](#)
> On Thu, Dec 24, 2020 at 4:51 AM Preslav Mihaylov ***@***.***> wrote: ***@***.**** requested changes on this pull request. ------------------------------ In issuetracker/gitlab/gitlab.go <[#118 (comment)](https://github.com/preslavmihaylov/todocheck/pull/118#discussion_r548521554)> : > func (it *IssueTracker) issueAPIOrigin() string { - tokens := common.RemoveEmptyTokens(strings.Split(it.Origin, "/")) - if !strings.HasPrefix(tokens[0], "http:") && !strings.HasPrefix(tokens[0], "https:") { - tokens = append([]string{"https:"}, tokens...) + + pattern := regexp.MustCompile(`^(?P<scheme>^(https?:))//(?P<domain>gitlab.([a-zA-Z0-9\-]+)?(.)?com)/(?P<owner>[a-zA-Z0-9\-]+)/(?P<repo>[a-zA-Z0-9-_.]+)/?$`) gitlab can be self-hosted. It should be able to match any host e.g. myorg.issuetracker.com ------------------------------ In issuetracker/gitlab/gitlab.go <[#118 (comment)](https://github.com/preslavmihaylov/todocheck/pull/118#discussion_r548521656)> : > func (it *IssueTracker) issueAPIOrigin() string { - tokens := common.RemoveEmptyTokens(strings.Split(it.Origin, "/")) - if !strings.HasPrefix(tokens[0], "http:") && !strings.HasPrefix(tokens[0], "https:") { - tokens = append([]string{"https:"}, tokens...) + + pattern := regexp.MustCompile(`^(?P<scheme>^(https?:))//(?P<domain>gitlab.([a-zA-Z0-9\-]+)?(.)?com)/(?P<owner>[a-zA-Z0-9\-]+)/(?P<repo>[a-zA-Z0-9-_.]+)/?$`) In this case, you expect someone to explicitly have gitlab in the hostname which can not always be the case. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#118 (review)](https://github.com/preslavmihaylov/todocheck/pull/118#pullrequestreview-558495018)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AEJAM7536KBXACKNHVIL673SWM2MTANCNFSM4UGEWEXA> .

Yep, that is correct

preslavmihaylov commented on issuetracker/gitlab/gitlab.go:
```diff
@@ -33,14 +34,19 @@ func (it *IssueTracker) taskURLFrom(taskID string) string {
-	if !strings.HasPrefix(tokens[0], "http:") && !strings.HasPrefix(tokens[0], "https:") {
-		tokens = append([]string{"https:"}, tokens...)
+
+	pattern := regexp.MustCompile(`^(?P<scheme>^(https?:))//(?P<domain>(gitlab.)?([a-zA-Z0-9\-]+)?(.)?com)/(?P<owner>[a-zA-Z0-9\-]+)/(?P<repo>[a-zA-Z0-9-_.]+)/?$`)
```
Currently, it is still more restrictive than it should be as it only allows `.com` domains. `myorg.net` won't work for example.

Also, the `(gitlab.)?` looks more like a hotfix and is not necessary at all

preslavmihaylov commented:
It seems that you have a merge conflict. Can you resolve it and we can merge the code
